# Assignment 5 - Neo4j

- Allan Simonsen
- Jean-Poul Leth-Møller
- Magda Aleksandra Wawrzak
- Nina Lisakowski

 # 1. Import of libraries

In [10]:
# import pandas for structuring the data
import pandas as pd

import numpy as np

from py2neo import Graph,Node,Relationship

# 2. import of dataset from CSV

In [3]:
game_dataset = pd.read_csv('../sample-data/data/Sample_Game_1/Sample_Game_1_RawEventsData.csv', delimiter=",")

In [4]:
game_dataset.head(5)

Team       Type   Subtype  Period  Start Frame  Start Time [s]  End Frame  \
0  Away  SET PIECE  KICK OFF       1            1            0.04          0   
1  Away       PASS       NaN       1            1            0.04          3   
2  Away       PASS       NaN       1            3            0.12         17   
3  Away       PASS       NaN       1           45            1.80         61   
4  Away       PASS       NaN       1           77            3.08         96   

   End Time [s]      From        To  Start X  Start Y  End X  End Y  
0          0.00  Player19       NaN      NaN      NaN    NaN    NaN  
1          0.12  Player19  Player21     0.45     0.39   0.55   0.43  
2          0.68  Player21  Player15     0.55     0.43   0.58   0.21  
3          2.44  Player15  Player19     0.55     0.19   0.45   0.31  
4          3.84  Player19  Player21     0.45     0.32   0.49   0.47

In [5]:
game_dataset.tail(5)

Team       Type       Subtype  Period  Start Frame  Start Time [s]  \
1740  Home       PASS           NaN       2       143361         5734.44   
1741  Home       PASS           NaN       2       143578         5743.12   
1742  Home  BALL LOST  INTERCEPTION       2       143598         5743.92   
1743  Away   RECOVERY       BLOCKED       2       143617         5744.68   
1744  Away   BALL OUT           NaN       2       143622         5744.88   

      End Frame  End Time [s]      From        To  Start X  Start Y  End X  \
1740     143483       5739.32  Player12  Player13     0.60     0.33   0.19   
1741     143593       5743.72  Player13   Player4     0.09     0.88   0.14   
1742     143618       5744.72   Player4       NaN     0.13     0.69   0.07   
1743     143617       5744.68  Player16       NaN     0.05     0.62    NaN   
1744     143630       5745.20  Player16       NaN     0.05     0.63   0.03   

      End Y  
1740   0.95  
1741   0.69  
1742   0.61  
1743    NaN  
1744   1.01

In [6]:
game_dataset.shape

(1745, 14)

In [7]:
game_dataset['From'].unique()

array(['Player19', 'Player21', 'Player15', 'Player22', 'Player17',
       'Player2', 'Player16', 'Player18', 'Player9', 'Player10',
       'Player20', 'Player5', 'Player6', 'Player8', 'Player25', 'Player1',
       'Player7', 'Player4', 'Player3', 'Player24', 'Player23',
       'Player11', 'Player12', 'Player27', 'Player13', 'Player26',
       'Player14', 'Player28'], dtype=object)

In [8]:
game_dataset['Type'].unique()

array(['SET PIECE', 'PASS', 'BALL LOST', 'RECOVERY', 'CHALLENGE',
       'BALL OUT', 'SHOT', 'FAULT RECEIVED', 'CARD'], dtype=object)

# 3. Connection to database

In [11]:
graph = Graph('bolt://localhost:7687', auth=('neo4j', 'root123'))

# 4. Create Player nodes

We decided to make all player as seperate nodes.  
furtermore we decided to make each event (PASS, SET KICK, Challenge, etc) as a new node with the time property and a relationship to and from related players.
The following picture show the setup with the PASS event.
![Graph Setup](GraphSetup.png)

In [8]:
for i in game_dataset['From'].unique():
    print(i)
    try:
        new_player = Node("player",name=i)
        tx=graph.begin()
        tx.create(new_player)
        graph.commit(tx)
    except:
        print('Player already Exist')
    
    


Player19
Player21
Player15
Player22
Player17
Player2
Player16
Player18
Player9
Player10
Player20
Player5
Player6
Player8
Player25
Player1
Player7
Player4
Player3
Player24
Player23
Player11
Player12
Player27
Player13
Player26
Player14
Player28


# 5. Create Relationships and Event Nodes

In [9]:
for i in game_dataset.itertuples():
    try:
        fromPlayer = graph.nodes.match("player", name=i.From).first()
        toPlayer = graph.nodes.match("player", name=i.To).first()
        new_Event_node = Node(i.Type, name=i.Type, subtype=i.Subtype, team=i.Team, period=i.Period, starttime=i._6, endtime=i._8)
        relationshipfrom = Relationship(fromPlayer, i.Type, new_Event_node)
        tx=graph.begin()
        tx.create(new_Event_node)
        tx.create(relationshipfrom)
        if(toPlayer is not None):
            relationshipTo = Relationship(new_Event_node, i.Type, toPlayer)
            tx.create(relationshipTo)
        graph.commit(tx)
        print("Event Done")
    except BaseException as E:
        print(E)

Pandas(Index=0, Team='Away', Type='SET PIECE', Subtype='KICK OFF', Period=1, _5=1, _6=0.04, _7=0, _8=0.0, From='Player19', To=nan, _11=nan, _12=nan, _13=nan, _14=nan)
type <class 'float'>
Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=1, Team='Away', Type='PASS', Subtype=nan, Period=1, _5=1, _6=0.04, _7=3, _8=0.12, From='Player19', To='Player21', _11=0.45, _12=0.39, _13=0.55, _14=0.43)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=2, Team='Away', Type='PASS', Subtype=nan, Period=1, _5=3, _6=0.12, _7=17, _8=0.68, From='Player21', To='Player15', _11=0.55, _12=0.43, _13=0.58, _14=0.21)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=3, Team='Away', Type='PASS', Subtype=nan, Period=1, _5=45, _6=1.8, _7=61, _8=2.44, From='Player15', To='Player19', _11=0.55, _12=0.19, _13=0.45, _14=0.31)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>

Hello3
Hello4
Event Done
Pandas(Index=37, Team='Away', Type='PASS', Subtype=nan, Period=1, _5=3780, _6=151.2, _7=3805, _8=152.2, From='Player21', To='Player20', _11=0.57, _12=0.53, _13=0.58, _14=0.73)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=38, Team='Away', Type='PASS', Subtype=nan, Period=1, _5=3829, _6=153.16, _7=3854, _8=154.16, From='Player20', To='Player22', _11=0.57, _12=0.76, _13=0.52, _14=0.98)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=39, Team='Away', Type='PASS', Subtype=nan, Period=1, _5=3882, _6=155.28, _7=3907, _8=156.28, From='Player22', To='Player17', _11=0.55, _12=0.95, _13=0.69, _14=0.87)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=40, Team='Away', Type='PASS', Subtype=nan, Period=1, _5=3930, _6=157.2, _7=3994, _8=159.76, From='Player17', To='Player25', _11=0.71, _12=0.86, _13=

Hello3
Hello4
Event Done
Pandas(Index=82, Team='Away', Type='BALL LOST', Subtype='HEAD', Period=1, _5=6779, _6=271.16, _7=6844, _8=273.76, From='Player18', To=nan, _11=0.53, _12=0.04, _13=0.57, _14=0.31)
type <class 'float'>
Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=83, Team='Home', Type='RECOVERY', Subtype=nan, Period=1, _5=6844, _6=273.76, _7=6844, _8=273.76, From='Player6', To=nan, _11=0.54, _12=0.3, _13=nan, _14=nan)
type <class 'float'>
Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=84, Team='Away', Type='CHALLENGE', Subtype='GROUND-WON', Period=1, _5=6863, _6=274.52, _7=6863, _8=274.52, From='Player21', To=nan, _11=0.53, _12=0.29, _13=nan, _14=nan)
type <class 'float'>
Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=85, Team='Away', Type='RECOVERY', Subtype='INTERCEPTION', Period=1, _5=6863, _6=274.52, _7=6863, _8=274.52, From='Player21', To=nan, _11=0.53, _12=0.29, _13=nan, _14=nan)
type <class 'fl

Pandas(Index=129, Team='Away', Type='PASS', Subtype=nan, Period=1, _5=9215, _6=368.6, _7=9255, _8=370.2, From='Player23', To='Player18', _11=0.06, _12=0.0, _13=0.22, _14=0.06)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=130, Team='Away', Type='PASS', Subtype=nan, Period=1, _5=9306, _6=372.24, _7=9337, _8=373.48, From='Player18', To='Player19', _11=0.18, _12=0.11, _13=0.24, _14=0.3)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=131, Team='Away', Type='BALL LOST', Subtype='INTERCEPTION', Period=1, _5=9337, _6=373.48, _7=9356, _8=374.24, From='Player19', To=nan, _11=0.24, _12=0.3, _13=0.24, _14=0.39)
type <class 'float'>
Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=132, Team='Home', Type='RECOVERY', Subtype='INTERCEPTION', Period=1, _5=9354, _6=374.16, _7=9354, _8=374.16, From='Player6', To=nan, _11=0.22, _12=0.36, _13=nan, _14=nan)
type <

Event Done
Pandas(Index=176, Team='Away', Type='BALL LOST', Subtype=nan, Period=1, _5=13766, _6=550.64, _7=13814, _8=552.56, From='Player15', To=nan, _11=0.64, _12=0.55, _13=0.76, _14=0.52)
type <class 'float'>
Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=177, Team='Home', Type='RECOVERY', Subtype=nan, Period=1, _5=13814, _6=552.56, _7=13814, _8=552.56, From='Player9', To=nan, _11=0.73, _12=0.56, _13=nan, _14=nan)
type <class 'float'>
Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=178, Team='Home', Type='CHALLENGE', Subtype='TACKLE-LOST', Period=1, _5=13843, _6=553.72, _7=13843, _8=553.72, From='Player9', To=nan, _11=0.8, _12=0.52, _13=nan, _14=nan)
type <class 'float'>
Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=179, Team='Home', Type='BALL LOST', Subtype='THEFT', Period=1, _5=13843, _6=553.72, _7=13843, _8=553.72, From='Player9', To=nan, _11=0.8, _12=0.52, _13=nan, _14=nan)
type <class 'float'>
Hello1


Event Done
Pandas(Index=222, Team='Home', Type='PASS', Subtype=nan, Period=1, _5=16162, _6=646.48, _7=16193, _8=647.72, From='Player2', To='Player3', _11=0.35, _12=0.79, _13=0.3, _14=0.53)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=223, Team='Home', Type='PASS', Subtype=nan, Period=1, _5=16223, _6=648.92, _7=16252, _8=650.08, From='Player3', To='Player6', _11=0.29, _12=0.57, _13=0.47, _14=0.67)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=224, Team='Home', Type='PASS', Subtype=nan, Period=1, _5=16252, _6=650.08, _7=16303, _8=652.12, From='Player6', To='Player4', _11=0.47, _12=0.67, _13=0.41, _14=0.22)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=225, Team='Home', Type='PASS', Subtype=nan, Period=1, _5=16332, _6=653.28, _7=16359, _8=654.36, From='Player4', To='Player10', _11=0.44, _12=0.22, _13=0.53, _

Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=270, Team='Home', Type='PASS', Subtype=nan, Period=1, _5=20362, _6=814.48, _7=20380, _8=815.2, From='Player7', To='Player4', _11=0.28, _12=0.14, _13=0.22, _14=0.1)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=271, Team='Home', Type='PASS', Subtype=nan, Period=1, _5=20471, _6=818.84, _7=20528, _8=821.12, From='Player4', To='Player9', _11=0.28, _12=0.03, _13=0.43, _14=0.51)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=272, Team='Home', Type='PASS', Subtype=nan, Period=1, _5=20552, _6=822.08, _7=20622, _8=824.88, From='Player9', To='Player1', _11=0.4, _12=0.53, _13=0.36, _14=0.91)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=273, Team='Home', Type='BALL LOST', Subtype='INTERCEPTION', Period=1, _5=20622, _6=824.88, _7=20671, _8=826.84,

Event Done
Pandas(Index=320, Team='Away', Type='CHALLENGE', Subtype='GROUND-WON', Period=1, _5=23812, _6=952.48, _7=23812, _8=952.48, From='Player21', To=nan, _11=0.29, _12=0.33, _13=nan, _14=nan)
type <class 'float'>
Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=321, Team='Home', Type='CHALLENGE', Subtype='GROUND-LOST', Period=1, _5=23812, _6=952.48, _7=23812, _8=952.48, From='Player8', To=nan, _11=0.29, _12=0.33, _13=nan, _14=nan)
type <class 'float'>
Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=322, Team='Away', Type='BALL LOST', Subtype='INTERCEPTION', Period=1, _5=23846, _6=953.84, _7=23862, _8=954.48, From='Player21', To=nan, _11=0.3, _12=0.38, _13=0.31, _14=0.45)
type <class 'float'>
Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=323, Team='Home', Type='RECOVERY', Subtype=nan, Period=1, _5=23860, _6=954.4, _7=23860, _8=954.4, From='Player9', To=nan, _11=0.32, _12=0.49, _13=nan, _14=nan)
type <class 

Hello3
Hello4
Event Done
Pandas(Index=374, Team='Away', Type='PASS', Subtype=nan, Period=1, _5=27657, _6=1106.28, _7=27673, _8=1106.92, From='Player21', To='Player19', _11=0.27, _12=0.8, _13=0.31, _14=0.72)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=375, Team='Away', Type='BALL LOST', Subtype='INTERCEPTION', Period=1, _5=27711, _6=1108.44, _7=27724, _8=1108.96, From='Player19', To=nan, _11=0.28, _12=0.72, _13=0.23, _14=0.84)
type <class 'float'>
Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=376, Team='Home', Type='RECOVERY', Subtype='INTERCEPTION', Period=1, _5=27724, _6=1108.96, _7=27724, _8=1108.96, From='Player5', To=nan, _11=0.22, _12=0.84, _13=nan, _14=nan)
type <class 'float'>
Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=377, Team='Home', Type='PASS', Subtype=nan, Period=1, _5=27724, _6=1108.96, _7=27746, _8=1109.84, From='Player5', To='Player1', _11=0.22, _12=0.

Event Done
Pandas(Index=425, Team='Home', Type='SHOT', Subtype='OFF TARGET-OUT', Period=1, _5=31190, _6=1247.6, _7=31235, _8=1249.4, From='Player7', To=nan, _11=0.75, _12=0.37, _13=1.05, _14=0.59)
type <class 'float'>
Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=426, Team='Away', Type='RECOVERY', Subtype=nan, Period=1, _5=31546, _6=1261.84, _7=31546, _8=1261.84, From='Player25', To=nan, _11=0.96, _12=0.62, _13=nan, _14=nan)
type <class 'float'>
Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=427, Team='Away', Type='PASS', Subtype='GOAL KICK', Period=1, _5=31546, _6=1261.84, _7=31597, _8=1263.88, From='Player25', To='Player17', _11=0.96, _12=0.62, _13=0.89, _14=0.88)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=428, Team='Away', Type='PASS', Subtype=nan, Period=1, _5=31709, _6=1268.36, _7=31733, _8=1269.32, From='Player17', To='Player20', _11=0.69, _12=0.88, _13=0.65, _14=0

Event Done
Pandas(Index=477, Team='Home', Type='RECOVERY', Subtype='INTERCEPTION', Period=1, _5=36370, _6=1454.8, _7=36370, _8=1454.8, From='Player8', To=nan, _11=0.56, _12=0.24, _13=nan, _14=nan)
type <class 'float'>
Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=478, Team='Home', Type='CHALLENGE', Subtype='GROUND-LOST', Period=1, _5=36410, _6=1456.4, _7=36410, _8=1456.4, From='Player8', To=nan, _11=0.59, _12=0.12, _13=nan, _14=nan)
type <class 'float'>
Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=479, Team='Home', Type='BALL LOST', Subtype='INTERCEPTION', Period=1, _5=36410, _6=1456.4, _7=36410, _8=1456.4, From='Player8', To=nan, _11=0.59, _12=0.12, _13=nan, _14=nan)
type <class 'float'>
Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=480, Team='Away', Type='CHALLENGE', Subtype='TACKLE-WON', Period=1, _5=36414, _6=1456.56, _7=36414, _8=1456.56, From='Player15', To=nan, _11=0.6, _12=0.12, _13=nan, _14=nan)


toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=531, Team='Home', Type='CHALLENGE', Subtype='GROUND-WON', Period=1, _5=43171, _6=1726.84, _7=43171, _8=1726.84, From='Player7', To=nan, _11=0.63, _12=0.98, _13=nan, _14=nan)
type <class 'float'>
Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=532, Team='Home', Type='PASS', Subtype=nan, Period=1, _5=43214, _6=1728.56, _7=43239, _8=1729.56, From='Player7', To='Player2', _11=0.56, _12=0.91, _13=0.45, _14=0.95)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=533, Team='Home', Type='BALL LOST', Subtype='INTERCEPTION', Period=1, _5=43270, _6=1730.8, _7=43324, _8=1732.96, From='Player2', To=nan, _11=0.45, _12=0.95, _13=0.75, _14=0.76)
type <class 'float'>
Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=534, Team='Home', Type='CHALLENGE', Subtype='GROUND', Period=1, _5=43307, _6=1732.28, _7=43307, _8=1732.2

type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=584, Team='Away', Type='PASS', Subtype=nan, Period=1, _5=47496, _6=1899.84, _7=47573, _8=1902.92, From='Player21', To='Player17', _11=0.39, _12=0.22, _13=0.46, _14=0.75)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=585, Team='Away', Type='PASS', Subtype=nan, Period=1, _5=47627, _6=1905.08, _7=47695, _8=1907.8, From='Player17', To='Player18', _11=0.43, _12=0.73, _13=0.23, _14=0.15)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=586, Team='Away', Type='PASS', Subtype=nan, Period=1, _5=47755, _6=1910.2, _7=47783, _8=1911.32, From='Player18', To='Player15', _11=0.31, _12=0.1, _13=0.42, _14=0.15)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=587, Team='Away', Type='PASS', Subtype=nan, Period=1, _5=47801, _

Event Done
Pandas(Index=624, Team='Home', Type='BALL LOST', Subtype='THEFT', Period=1, _5=51224, _6=2048.96, _7=51224, _8=2048.96, From='Player9', To=nan, _11=0.56, _12=0.88, _13=nan, _14=nan)
type <class 'float'>
Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=625, Team='Away', Type='CHALLENGE', Subtype='TACKLE-WON', Period=1, _5=51225, _6=2049.0, _7=51225, _8=2049.0, From='Player20', To=nan, _11=0.57, _12=0.87, _13=nan, _14=nan)
type <class 'float'>
Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=626, Team='Away', Type='RECOVERY', Subtype='THEFT', Period=1, _5=51225, _6=2049.0, _7=51225, _8=2049.0, From='Player20', To=nan, _11=0.57, _12=0.87, _13=nan, _14=nan)
type <class 'float'>
Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=627, Team='Away', Type='PASS', Subtype=nan, Period=1, _5=51225, _6=2049.0, _7=51226, _8=2049.04, From='Player20', To='Player19', _11=0.57, _12=0.87, _13=0.55, _14=0.89)
type <class 'str

Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=677, Team='Away', Type='PASS', Subtype=nan, Period=1, _5=54381, _6=2175.24, _7=54411, _8=2176.44, From='Player21', To='Player19', _11=0.69, _12=0.92, _13=0.68, _14=0.75)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=678, Team='Away', Type='PASS', Subtype=nan, Period=1, _5=54476, _6=2179.04, _7=54506, _8=2180.24, From='Player19', To='Player23', _11=0.64, _12=0.69, _13=0.52, _14=0.55)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=679, Team='Home', Type='CHALLENGE', Subtype='GROUND-LOST', Period=1, _5=54505, _6=2180.2, _7=54505, _8=2180.2, From='Player3', To=nan, _11=0.51, _12=0.56, _13=nan, _14=nan)
type <class 'float'>
Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=680, Team='Away', Type='CHALLENGE', Subtype='GROUND-WON', Period=1, _5=54506, _6=2180.24, _7=54506, _8=218

toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=728, Team='Away', Type='BALL LOST', Subtype=nan, Period=1, _5=57624, _6=2304.96, _7=57669, _8=2306.76, From='Player16', To=nan, _11=0.95, _12=0.69, _13=0.81, _14=0.74)
type <class 'float'>
Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=729, Team='Home', Type='RECOVERY', Subtype=nan, Period=1, _5=57668, _6=2306.72, _7=57668, _8=2306.72, From='Player7', To=nan, _11=0.81, _12=0.73, _13=nan, _14=nan)
type <class 'float'>
Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=730, Team='Home', Type='SHOT', Subtype='OFF TARGET-OUT', Period=1, _5=57668, _6=2306.72, _7=57740, _8=2309.6, From='Player7', To=nan, _11=0.81, _12=0.73, _13=1.02, _14=0.31)
type <class 'float'>
Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=731, Team='Away', Type='RECOVERY', Subtype=nan, Period=1, _5=59053, _6=2362.12, _7=59053, _8=2362.12, From='Player25', To=nan, _11=

Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=778, Team='Away', Type='CHALLENGE', Subtype='GROUND-WON', Period=1, _5=63407, _6=2536.28, _7=63407, _8=2536.28, From='Player18', To=nan, _11=0.51, _12=0.47, _13=nan, _14=nan)
type <class 'float'>
Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=779, Team='Home', Type='CHALLENGE', Subtype='GROUND-ADVANTAGE-LOST', Period=1, _5=63428, _6=2537.12, _7=63428, _8=2537.12, From='Player5', To=nan, _11=0.52, _12=0.43, _13=nan, _14=nan)
type <class 'float'>
Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=780, Team='Away', Type='CHALLENGE', Subtype='GROUND-FAULT-WON', Period=1, _5=63469, _6=2538.76, _7=63469, _8=2538.76, From='Player18', To=nan, _11=0.53, _12=0.35, _13=nan, _14=nan)
type <class 'float'>
Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=781, Team='Away', Type='FAULT RECEIVED', Subtype=nan, Period=1, _5=63477, _6=2539.08, _7=

Event Done
Pandas(Index=832, Team='Home', Type='PASS', Subtype=nan, Period=1, _5=66702, _6=2668.08, _7=66729, _8=2669.16, From='Player7', To='Player5', _11=0.47, _12=1.01, _13=0.54, _14=0.85)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=833, Team='Home', Type='PASS', Subtype=nan, Period=1, _5=66729, _6=2669.16, _7=66747, _8=2669.88, From='Player5', To='Player9', _11=0.54, _12=0.85, _13=0.61, _14=0.92)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=834, Team='Home', Type='CHALLENGE', Subtype='GROUND-WON', Period=1, _5=66747, _6=2669.88, _7=66747, _8=2669.88, From='Player9', To=nan, _11=0.61, _12=0.92, _13=nan, _14=nan)
type <class 'float'>
Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=835, Team='Home', Type='BALL LOST', Subtype=nan, Period=1, _5=66747, _6=2669.88, _7=66803, _8=2672.12, From='Player9', To=nan, _11=0.61, _12=0.92, _13=0.71, _

Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=882, Team='Away', Type='PASS', Subtype=nan, Period=1, _5=69269, _6=2770.76, _7=69288, _8=2771.52, From='Player17', To='Player25', _11=0.93, _12=0.33, _13=0.98, _14=0.47)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=883, Team='Away', Type='PASS', Subtype=nan, Period=1, _5=69357, _6=2774.28, _7=69397, _8=2775.88, From='Player25', To='Player19', _11=0.9, _12=0.52, _13=0.82, _14=0.67)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=884, Team='Away', Type='PASS', Subtype=nan, Period=1, _5=69495, _6=2779.8, _7=69538, _8=2781.52, From='Player19', To='Player24', _11=0.68, _12=0.66, _13=0.51, _14=0.26)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=885, Team='Away', Type='PASS', Subtype=nan, Period=1, _5=69563, _6=2782.52, _7=69588, _8=2783.52, 

Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=935, Team='Away', Type='PASS', Subtype=nan, Period=2, _5=74979, _6=2999.16, _7=74998, _8=2999.92, From='Player19', To='Player17', _11=0.56, _12=0.87, _13=0.57, _14=1.0)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=936, Team='Away', Type='PASS', Subtype=nan, Period=2, _5=75022, _6=3000.88, _7=75042, _8=3001.68, From='Player17', To='Player19', _11=0.56, _12=0.97, _13=0.52, _14=0.89)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=937, Team='Away', Type='PASS', Subtype=nan, Period=2, _5=75101, _6=3004.04, _7=75176, _8=3007.04, From='Player19', To='Player24', _11=0.54, _12=0.82, _13=0.78, _14=0.7)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=938, Team='Away', Type='PASS', Subtype=nan, Period=2, _5=75232, _6=3009.28, _7=7526

Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=987, Team='Home', Type='SET PIECE', Subtype='THROW IN', Period=2, _5=79677, _6=3187.08, _7=79677, _8=3187.08, From='Player4', To=nan, _11=nan, _12=nan, _13=nan, _14=nan)
type <class 'float'>
Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=988, Team='Home', Type='PASS', Subtype=nan, Period=2, _5=79677, _6=3187.08, _7=79699, _8=3187.96, From='Player4', To='Player6', _11=0.84, _12=1.01, _13=0.77, _14=0.94)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=989, Team='Home', Type='PASS', Subtype=nan, Period=2, _5=79699, _6=3187.96, _7=79724, _8=3188.96, From='Player6', To='Player4', _11=0.77, _12=0.94, _13=0.85, _14=0.98)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=990, Team='Home', Type='BALL LOST', Subtype='INTERCEPTION', Period=2, _5=79724, _6=3188.96, _7=79780, _8=3191.2, 

Hello3
Event Done
Pandas(Index=1039, Team='Away', Type='PASS', Subtype=nan, Period=2, _5=83306, _6=3332.24, _7=83328, _8=3333.12, From='Player19', To='Player21', _11=0.65, _12=0.82, _13=0.58, _14=0.77)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=1040, Team='Away', Type='PASS', Subtype=nan, Period=2, _5=83425, _6=3337.0, _7=83455, _8=3338.2, From='Player21', To='Player18', _11=0.65, _12=0.54, _13=0.64, _14=0.32)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=1041, Team='Away', Type='PASS', Subtype=nan, Period=2, _5=83485, _6=3339.4, _7=83510, _8=3340.4, From='Player18', To='Player22', _11=0.66, _12=0.27, _13=0.72, _14=0.06)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=1042, Team='Away', Type='PASS', Subtype=nan, Period=2, _5=83575, _6=3343.0, _7=83600, _8=3344.0, From='Player22', To='Player21', _11=0.71, 

Event Done
Pandas(Index=1093, Team='Home', Type='BALL OUT', Subtype='WOODWORK-GOAL', Period=2, _5=87470, _6=3498.8, _7=87485, _8=3499.4, From='Player5', To=nan, _11=0.97, _12=0.48, _13=1.01, _14=0.54)
type <class 'float'>
Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=1094, Team='Home', Type='SET PIECE', Subtype='KICK OFF', Period=2, _5=88792, _6=3551.68, _7=88792, _8=3551.68, From='Player10', To=nan, _11=nan, _12=nan, _13=nan, _14=nan)
type <class 'float'>
Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=1095, Team='Home', Type='PASS', Subtype=nan, Period=2, _5=88792, _6=3551.68, _7=88820, _8=3552.8, From='Player10', To='Player12', _11=0.49, _12=0.5, _13=0.56, _14=0.41)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=1096, Team='Home', Type='PASS', Subtype=nan, Period=2, _5=88820, _6=3552.8, _7=88847, _8=3553.88, From='Player12', To='Player6', _11=0.56, _12=0.41, _13=0.56, _14=

Event Done
Pandas(Index=1145, Team='Home', Type='CHALLENGE', Subtype='AERIAL-LOST', Period=2, _5=93670, _6=3746.8, _7=93670, _8=3746.8, From='Player9', To=nan, _11=0.35, _12=0.55, _13=nan, _14=nan)
type <class 'float'>
Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=1146, Team='Away', Type='CHALLENGE', Subtype='AERIAL-WON', Period=2, _5=93672, _6=3746.88, _7=93672, _8=3746.88, From='Player16', To=nan, _11=0.34, _12=0.55, _13=nan, _14=nan)
type <class 'float'>
Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=1147, Team='Away', Type='RECOVERY', Subtype='INTERCEPTION', Period=2, _5=93672, _6=3746.88, _7=93672, _8=3746.88, From='Player16', To=nan, _11=0.34, _12=0.55, _13=nan, _14=nan)
type <class 'float'>
Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=1148, Team='Away', Type='BALL LOST', Subtype='HEAD', Period=2, _5=93672, _6=3746.88, _7=93719, _8=3748.76, From='Player16', To=nan, _11=0.34, _12=0.55, _13=0.41, _14=0

Hello4
Event Done
Pandas(Index=1199, Team='Home', Type='PASS', Subtype=nan, Period=2, _5=98407, _6=3936.28, _7=98432, _8=3937.28, From='Player4', To='Player8', _11=0.46, _12=0.88, _13=0.41, _14=0.98)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=1200, Team='Home', Type='PASS', Subtype=nan, Period=2, _5=98443, _6=3937.72, _7=98473, _8=3938.92, From='Player8', To='Player6', _11=0.42, _12=0.98, _13=0.48, _14=0.85)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=1201, Team='Home', Type='PASS', Subtype=nan, Period=2, _5=98530, _6=3941.2, _7=98576, _8=3943.04, From='Player6', To='Player12', _11=0.49, _12=0.86, _13=0.44, _14=0.54)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=1202, Team='Home', Type='PASS', Subtype=nan, Period=2, _5=98615, _6=3944.6, _7=98663, _8=3946.52, From='Player12', To='Player5', _11=0.4, _12

Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=1253, Team='Away', Type='RECOVERY', Subtype='INTERCEPTION', Period=2, _5=103917, _6=4156.68, _7=103917, _8=4156.68, From='Player15', To=nan, _11=0.16, _12=0.71, _13=nan, _14=nan)
type <class 'float'>
Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=1254, Team='Away', Type='PASS', Subtype=nan, Period=2, _5=103917, _6=4156.68, _7=103957, _8=4158.28, From='Player15', To='Player17', _11=0.16, _12=0.71, _13=0.12, _14=0.74)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=1255, Team='Home', Type='BALL LOST', Subtype='INTERCEPTION', Period=2, _5=103918, _6=4156.72, _7=103918, _8=4156.72, From='Player8', To=nan, _11=0.18, _12=0.73, _13=0.14, _14=0.61)
type <class 'float'>
Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=1256, Team='Away', Type='BALL LOST', Subtype='CLEARANCE', Period=2, _5=103957, _6=41

Event Done
Pandas(Index=1306, Team='Away', Type='RECOVERY', Subtype='INTERCEPTION', Period=2, _5=107708, _6=4308.32, _7=107708, _8=4308.32, From='Player16', To=nan, _11=0.15, _12=0.13, _13=nan, _14=nan)
type <class 'float'>
Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=1307, Team='Away', Type='PASS', Subtype=nan, Period=2, _5=107752, _6=4310.08, _7=107777, _8=4311.08, From='Player16', To='Player15', _11=0.11, _12=0.19, _13=0.08, _14=0.33)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=1308, Team='Away', Type='PASS', Subtype=nan, Period=2, _5=107787, _6=4311.48, _7=107802, _8=4312.08, From='Player15', To='Player16', _11=0.07, _12=0.32, _13=0.12, _14=0.24)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=1309, Team='Away', Type='BALL LOST', Subtype='INTERCEPTION', Period=2, _5=107805, _6=4312.2, _7=107830, _8=4313.2, From='Player16', To=nan, _11

Event Done
Pandas(Index=1356, Team='Home', Type='PASS', Subtype=nan, Period=2, _5=113294, _6=4531.76, _7=113314, _8=4532.56, From='Player3', To='Player12', _11=0.69, _12=0.65, _13=0.61, _14=0.75)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=1357, Team='Home', Type='PASS', Subtype=nan, Period=2, _5=113320, _6=4532.8, _7=113347, _8=4533.88, From='Player12', To='Player3', _11=0.61, _12=0.75, _13=0.73, _14=0.64)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=1358, Team='Home', Type='BALL LOST', Subtype='INTERCEPTION', Period=2, _5=113390, _6=4535.6, _7=113448, _8=4537.92, From='Player3', To=nan, _11=0.77, _12=0.56, _13=0.47, _14=0.3)
type <class 'float'>
Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=1359, Team='Away', Type='CHALLENGE', Subtype='AERIAL-WON', Period=2, _5=113447, _6=4537.88, _7=113447, _8=4537.88, From='Player20', To=nan, _11=0.

Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=1408, Team='Home', Type='PASS', Subtype=nan, Period=2, _5=116206, _6=4648.24, _7=116236, _8=4649.44, From='Player7', To='Player13', _11=0.45, _12=0.19, _13=0.33, _14=0.05)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=1409, Team='Home', Type='PASS', Subtype=nan, Period=2, _5=116246, _6=4649.84, _7=116274, _8=4650.96, From='Player13', To='Player7', _11=0.33, _12=0.04, _13=0.41, _14=0.11)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=1410, Team='Home', Type='PASS', Subtype=nan, Period=2, _5=116289, _6=4651.56, _7=116301, _8=4652.04, From='Player7', To='Player10', _11=0.41, _12=0.1, _13=0.34, _14=0.19)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=1411, Team='Home', Type='BALL LOST', Subtype='REFEREE HIT', Period=2, _5=116

Hello3
Event Done
Pandas(Index=1459, Team='Away', Type='RECOVERY', Subtype=nan, Period=2, _5=118960, _6=4758.4, _7=118960, _8=4758.4, From='Player25', To=nan, _11=0.01, _12=0.44, _13=nan, _14=nan)
type <class 'float'>
Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=1460, Team='Away', Type='PASS', Subtype='GOAL KICK', Period=2, _5=119733, _6=4789.32, _7=119768, _8=4790.72, From='Player25', To='Player15', _11=0.04, _12=0.5, _13=0.1, _14=0.83)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=1461, Team='Away', Type='PASS', Subtype=nan, Period=2, _5=119810, _6=4792.4, _7=119842, _8=4793.68, From='Player15', To='Player21', _11=0.12, _12=0.82, _13=0.27, _14=0.73)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=1462, Team='Away', Type='PASS', Subtype=nan, Period=2, _5=119860, _6=4794.4, _7=119885, _8=4795.4, From='Player21', To='Player17', _11=0.28, _12

Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=1513, Team='Away', Type='RECOVERY', Subtype='INTERCEPTION', Period=2, _5=123920, _6=4956.8, _7=123920, _8=4956.8, From='Player16', To=nan, _11=0.36, _12=0.37, _13=nan, _14=nan)
type <class 'float'>
Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=1514, Team='Away', Type='PASS', Subtype=nan, Period=2, _5=124005, _6=4960.2, _7=124045, _8=4961.8, From='Player16', To='Player25', _11=0.15, _12=0.25, _13=0.01, _14=0.32)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=1515, Team='Away', Type='BALL LOST', Subtype='INTERCEPTION', Period=2, _5=124051, _6=4962.04, _7=124146, _8=4965.84, From='Player25', To=nan, _11=0.01, _12=0.32, _13=0.45, _14=0.43)
type <class 'float'>
Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=1516, Team='Home', Type='RECOVERY', Subtype='INTERCEPTION', Period=2, _5=124147, _6=4965.88, _7=124147

Event Done
Pandas(Index=1566, Team='Home', Type='PASS', Subtype='HEAD', Period=2, _5=127891, _6=5115.64, _7=127911, _8=5116.44, From='Player3', To='Player12', _11=0.85, _12=0.55, _13=0.81, _14=0.6)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=1567, Team='Home', Type='PASS', Subtype=nan, Period=2, _5=127919, _6=5116.76, _7=127954, _8=5118.16, From='Player12', To='Player4', _11=0.82, _12=0.61, _13=0.89, _14=0.75)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=1568, Team='Home', Type='PASS', Subtype=nan, Period=2, _5=127959, _6=5118.36, _7=128009, _8=5120.36, From='Player4', To='Player8', _11=0.89, _12=0.77, _13=0.66, _14=0.94)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=1569, Team='Home', Type='PASS', Subtype=nan, Period=2, _5=128058, _6=5122.32, _7=128088, _8=5123.52, From='Player8', To='Player13', _11=0.

Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=1610, Team='Home', Type='PASS', Subtype=nan, Period=2, _5=131256, _6=5250.24, _7=131319, _8=5252.76, From='Player3', To='Player7', _11=0.54, _12=0.78, _13=0.45, _14=0.15)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=1611, Team='Home', Type='PASS', Subtype=nan, Period=2, _5=131399, _6=5255.96, _7=131424, _8=5256.96, From='Player7', To='Player14', _11=0.39, _12=0.13, _13=0.31, _14=0.06)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=1612, Team='Home', Type='PASS', Subtype=nan, Period=2, _5=131429, _6=5257.16, _7=131449, _8=5257.96, From='Player14', To='Player7', _11=0.31, _12=0.06, _13=0.37, _14=0.14)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=1613, Team='Home', Type='PASS', Subtype=nan, Period=2, _5=131466, _6=5258.64

Hello4
Event Done
Pandas(Index=1659, Team='Away', Type='PASS', Subtype=nan, Period=2, _5=135494, _6=5419.76, _7=135524, _8=5420.96, From='Player17', To='Player28', _11=0.81, _12=0.79, _13=0.75, _14=0.66)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=1660, Team='Away', Type='BALL LOST', Subtype='INTERCEPTION', Period=2, _5=135531, _6=5421.24, _7=135561, _8=5422.44, From='Player28', To=nan, _11=0.75, _12=0.67, _13=0.87, _14=0.32)
type <class 'float'>
Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=1661, Team='Home', Type='RECOVERY', Subtype='INTERCEPTION', Period=2, _5=135566, _6=5422.64, _7=135566, _8=5422.64, From='Player7', To=nan, _11=0.88, _12=0.35, _13=nan, _14=nan)
type <class 'float'>
Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=1662, Team='Home', Type='PASS', Subtype=nan, Period=2, _5=135585, _6=5423.4, _7=135595, _8=5423.8, From='Player7', To='Player14', _11=0.87, _

Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=1713, Team='Away', Type='RECOVERY', Subtype='INTERCEPTION', Period=2, _5=140675, _6=5627.0, _7=140675, _8=5627.0, From='Player20', To=nan, _11=0.41, _12=0.15, _13=nan, _14=nan)
type <class 'float'>
Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=1714, Team='Away', Type='PASS', Subtype=nan, Period=2, _5=140729, _6=5629.16, _7=140744, _8=5629.76, From='Player20', To='Player28', _11=0.42, _12=0.04, _13=0.5, _14=0.15)
type <class 'str'>
Hello1
toPlayer <class 'py2neo.data.Node'>
HEllo2
Hello3
Hello4
Event Done
Pandas(Index=1715, Team='Home', Type='CHALLENGE', Subtype='TACKLE-FAULT-LOST', Period=2, _5=140893, _6=5635.72, _7=140893, _8=5635.72, From='Player14', To=nan, _11=0.56, _12=0.41, _13=nan, _14=nan)
type <class 'float'>
Hello1
toPlayer <class 'NoneType'>
HEllo2
Hello3
Event Done
Pandas(Index=1716, Team='Away', Type='CHALLENGE', Subtype='TACKLE-FAULT-WON', Period=2, _5=140912, _6=5636.48, 

# 6. Getting Data from the graph

## 6.1 Who is the most active player (in terms of passing and receiving the ball)?

We count the amount of PASS relationship to and from each player node

In [12]:
activity_result=graph.run("MATCH(player:player)-[event:PASS]-(:PASS) WITH player, COUNT(event) AS num ORDER BY num DESC RETURN player.name,num")
df = pd.DataFrame(activity_result)
df

0    1
0   Player21  135
1   Player12  106
2    Player7  102
3    Player4   89
4   Player19   85
5    Player2   84
6    Player6   83
7   Player17   80
8    Player8   76
9    Player3   74
10  Player16   71
11  Player10   70
12   Player5   69
13  Player20   66
14  Player22   61
15  Player15   57
16  Player18   55
17   Player9   44
18  Player25   27
19  Player23   27
20  Player11   25
21  Player27   25
22  Player14   20
23   Player1   17
24  Player28   17
25  Player13   15
26  Player24   14
27  Player26    4

#### Answer: 
Player21 is the most active player in terms of passing and recieving the ball with a total of 135 PASS events related to him/her.

## 6.2 who has had a central role in the match?

we count the total amount of relationships each players have to find the most active player.  
one could argue that some events are worth more than others, however in this case we count the equally.

In [20]:
activity_result=graph.run("MATCH(player:player)-[event]-() WITH player, COUNT(event) AS num ORDER BY num DESC RETURN player.name,num")
df = pd.DataFrame(activity_result)
df

0    1
0   Player21  172
1   Player19  149
2    Player7  149
3    Player8  137
4   Player12  135
5   Player17  134
6    Player6  130
7    Player2  129
8    Player4  123
9    Player3  117
10  Player16  116
11  Player15  109
12  Player22  105
13  Player10  103
14   Player5   98
15  Player18   95
16  Player20   94
17   Player9   81
18  Player25   62
19  Player11   62
20  Player23   46
21  Player27   37
22   Player1   37
23  Player14   34
24  Player28   28
25  Player24   27
26  Player13   19
27  Player26   16

#### Answer
We see that player21 has the most relationships, we therefore conclude that he/she had the biggest impact on the match

## 6.3 which players have attempted to score?

We count the amount of SHOT reltionships for each player

In [13]:
activity_result=graph.run("MATCH(player:player)-[event:SHOT]-(:SHOT) WITH player, COUNT(event) AS num ORDER BY num DESC RETURN player.name,num")
df = pd.DataFrame(activity_result)
df


0  1
0    Player9  4
1   Player10  4
2    Player8  4
3    Player7  2
4    Player5  2
5   Player21  1
6   Player24  1
7   Player19  1
8   Player12  1
9   Player27  1
10  Player14  1
11  Player23  1
12  Player28  1

In [14]:
activity_result=graph.run("MATCH(player:player)-[event:SHOT]-(eventNode:SHOT) WHERE eventNode.subtype CONTAINS 'GOAL' WITH player, COUNT(event) AS num ORDER BY num DESC RETURN player.name,num")
df = pd.DataFrame(activity_result)
df

0  1
0   Player9  2
1  Player10  1

#### Answer
Quite a few players have tried to score goals, player9 and player10 both succeeded in goals and the match ended 3-0

## 6.4  which team has kept the ball longer?

We add up the start minus the end time and sum them together for each team

In [24]:
activity_result_away=graph.run("MATCH (p:player)-[:PASS]->(e:PASS)-[:PASS]->(p2:player) WHERE e.team = 'Away' RETURN sum(toInteger(e.endtime) - toInteger(e.starttime))")
activity_result_away



sum(toInteger(e.endtime) - toInteger(e.starttime)) 
----------------------------------------------------
                                                491

In [26]:
activity_result_Home=graph.run("MATCH (p:player)-[:PASS]->(e:PASS)-[:PASS]->(p2:player) WHERE e.team = 'Home' RETURN sum(toInteger(e.endtime) - toInteger(e.starttime))")
activity_result_Home

sum(toInteger(e.endtime) - toInteger(e.starttime)) 
----------------------------------------------------
                                                613

#### Answer
We can see that the home team has had to ball for more time than the away team

## 6.5  is there any close ‘societies’ between players (passing the ball to each other)?

we count the amount of PASS relationships between two players

In [16]:
activity_result=graph.run("MATCH(player:player)-[event:PASS]-(:PASS)-[r2:PASS]-(toPlayer:player)WITH player, toPlayer,COUNT(event) AS num ORDER BY num DESC RETURN player.name, toPlayer.name, num")
df = pd.DataFrame(activity_result)
df.head(10)


0         1   2
0   Player7   Player2  25
1   Player2   Player7  25
2  Player21  Player19  24
3  Player19  Player21  24
4   Player4   Player8  23
5   Player8   Player4  23
6   Player5   Player7  19
7   Player7   Player5  19
8  Player22  Player21  18
9  Player21  Player22  18

#### Answer
We can see that player7-player2, player21-19 and player4-player8 has the highest amount of passing the ball to eachother

## 6.6  how close is the connection between two specific players?

Since players dont have relationship directly to eachother we will count the amount of PASS event nodes there is between two players

In [13]:
player1= "Player21"
player2= "Player15"
activity_result=graph.run("MATCH(player:player{name:'"+ player1 +"'})-[event:PASS]-(:PASS)-[r2:PASS]-(toPlayer:player{name:'"+ player2 +"'})WITH player, toPlayer,COUNT(event) AS num ORDER BY num DESC RETURN player.name, toPlayer.name, num")
df = pd.DataFrame(activity_result)
df.head(10)


0         1   2
0  Player21  Player15  14

#### Answer
We can see that there are 14 connection between 2 specific players, in this case player21 and player15

_______________________________________________________

In [7]:
graph.evaluate("MATCH (n) DETACH DELETE n")